src/executor/graph_executor.cc

```diff
@@ -1211,63 +1212,53 @@ void GraphExecutor::InitOpSegs() {
 
 
 void GraphExecutor::BulkTrainingOpSegs(size_t total_num_nodes) {
-  // The maximum number of node in a segment executed in bulk
-  size_t num_nodes_threshold = dmlc::GetEnv("MXNET_EXEC_BULK_EXEC_MAX_NODE_TRAIN", 15);
+  // The maximum number of nodes in a segment executed in bulk (excluding variables) in fwd pass.
+  size_t segment_num_nodes_threshold_fwd = Imperative::BulkExecMaxNodeTrainFwd();
+  // The maximum number of nodes in a segment executed in bulk (excluding variables) in bwd pass.
+  size_t segment_num_nodes_threshold_bwd = Imperative::BulkExecMaxNodeTrainBwd();
 
   // create forward segments for training
   size_t topo_start = 0;
+  size_t segment_node_count = 0;
   for (size_t nid = 0; nid < num_forward_nodes_; nid++) {
```

Comment:
The inner loops for forward and backward bulking are very similar. Do you think it's better to refactor them as functions?

Reply:
I think having a single function to bulk forward and backward is a good idea, such as:
```
void GraphExecutor::BulkTrainingOpSegsOverRange(size_t from_node, size_t up_to_node, size_t max_nodes_per_segment) { ... }
```
The only issue to work out is if there's a good reason why in forward we have:
```
bool ignore_node = node->is_variable();
```
while in backward we have the more complicated:
```
bool ignore_node = node->is_variable() || op_node.skip_exec_node || op_node.exec == nullptr; 
```
Do you know by any chance?  The goal would be to unify these two as much as possible.

Reply:
Should be fine to always use 
```
bool ignore_node = node->is_variable() || op_node.skip_exec_node || op_node.exec == nullptr; 
```
`skip_exec_node` defaults to False. Is it set to True if the node should be skipped during execution. This is usually for elemwise_add op during backward pass. It is set when the elemwise_add op is skipped, and the req for the corresponding entry at the previous op is set to `kAddTo`. For forward pass, `skip_exec_node` should always be False. 
https://github.com/apache/incubator-mxnet/blob/992c3c0dd90c0723de6934e826a49bad6569eeac/src/executor/exec_pass.h#L162

It should also be ok to check `op_node.exec` for forward ops, as non variable nodes are assigned with exec: https://github.com/apache/incubator-mxnet/blob/master/src/executor/graph_executor.cc#L1089-L1094

Reply:
Your comments were very helpful.  Based on the encouragement, I consolidated the routines for bulking, going a little farther than what you suggested by including the forward inference bulking as well.  CI is offline right now.